### ISSUE TITLE
Allow customization of cache key generation

### ISSUE THREAD

**Issue by zanieb** [karpetrosyan/hishel#75]
Currently, `CacheTransport.handle_request` calls `generate_key` to create a cache key for a request. With this design, the key cannot be modified by a user of the library.

I'd like for `generate_key` to be moved to the `Controller` so that a controller subclass can implement a custom extension of key based on the request.

**Comment by karosis88** [comment]
> CacheTransport.handle_request

Thank you so much for being involved in this library🙏☺️ Do we want to move that function into the controller class or just offer it as an argument for storage?

**Comment by zanieb** [comment]
It feels more natural on the controller to me, but I just started using the library :) I don't see different storage types using different keys so I wouldn't expect to need to create a storage class to change this.

**Comment by karosis88** [comment]
At this point in time, module "_controller.py" holds only the RFC 9111 implementation. I think it's better to add a storage argument for this feature, as this seems to be just a storage feature and not controller-related.

Potentially, the new storage signature can look like this:

```python
def generate_key(request: Request) -> str:
    'default key generation'
    return "0"

class FileStorage:
    def __init__(
        self,
        serializer: tp.Optional[BaseSerializer] = None,
        base_path: tp.Optional[Path] = None,
        ttl: tp.Optional[int] = None,
        key_generator: tp.Union[tp.Callable[[Request], str], None] = None
    ) -> None:
        self._key_generator = key_generator or generate_key
```

What are your thoughts on this?

**Comment by zanieb** [comment]
I still feel weird specifying cache key generation on the storage type — ideally any storage type would be usable with a different cache key implementation.

There's a [section of the RFC](https://www.rfc-editor.org/rfc/rfc9111.html#name-overview-of-cache-operation) which briefly discusses cache key generation and usage of the URI as the key. It seems like they include cache key generation and extension of cache keys as part of the specification. Personally, I'd put this in the `Controller` rather than in a utility module (where it is now). I can see an argument for including it in the `Storage` abstraction since it already needs to be aware of `Request` objects, but it doesn't feel quite right.

I'm in no rush for this, so we can always see if someone else chimes in.

**Comment by karpetrosyan** [comment]
I think, if we think about the design, both places are correct for cache key implementation. 
The one thing is that, depending on where we put it, there are pros and cons for some particular use cases.

For example, putting it in the `Controller` class allows us to use a single storage with different controllers and different cache key implementations for each of them.

Putting it into the controller seems like a more flexible solution to me. 
The example above is very hard to solve when the cache key implementation is in storage. 
We will consider creating two `storages`, which will cause race conditions.

After reviewing it once more, I am okay with putting it in the controller class.
In some rare cases, we will see that the cache key in the storage class is less flexible than in the controller class.

**Comment by parkerhancock** [comment]
My vote is for baking this into the transport. See #130.

**Comment by karpetrosyan** [comment]
Resolved in #130
